examples/word_language_model/learner.py

```diff
@@ -0,0 +1,92 @@
+import inferno
+import torch
+from torch.autograd import Variable
+from sklearn.metrics import f1_score
+
+class Learner(inferno.NeuralNet):
+
+    def __init__(self,
+                 criterion=torch.nn.CrossEntropyLoss,
+                 clip=0.25,
+                 lr=20,
+                 ntokens=10000,
+                 *args, **kwargs):
+        self.clip = clip
+        self.ntokens = ntokens
+        super(Learner, self).__init__(criterion=criterion, lr=lr, *args, **kwargs)
+
+    def repackage_hidden(self, h):
+        """Wraps hidden states in new Variables, to detach them from their history."""
+        if isinstance(h, Variable):
+            v = Variable(h.data)
+            return v.cuda() if self.use_cuda else v
+        else:
+            return tuple(self.repackage_hidden(v) for v in h)
+
+    def on_train_begin(self, net, *args, **kwargs):
+        self.hidden = self.module_.init_hidden(self.batch_size)
+        self.module_.cuda()
+
+    def train_step(self, X, y, _):
+        self.module_.train()
+
+        self.hidden = self.repackage_hidden(self.hidden)
+        self.module_.zero_grad()
+
+        output, self.hidden = self.module_(X, self.hidden)
+        y_pred = output.view(-1, self.ntokens)
+
+        loss = self.get_loss(y_pred, y)
+        loss.backward()
+
+        torch.nn.utils.clip_grad_norm(self.module_.parameters(), self.clip)
+        for p in self.module_.parameters():
+            p.data.add_(-self.lr, p.grad.data)
+
+        return loss
+
+    def validation_step(self, X, y):
+        self.module_.eval()
+
+        output, self.hidden = self.module_(X, self.hidden)
+        output_flat = output.view(-1, self.ntokens)
+
+        return self.get_loss(output_flat, y)
+
+    def evaluation_step(self, X):
+        self.module_.eval()
+
+        hidden = self.module_.init_hidden(X.size(1))
+        output, hidden = self.module_(X, hidden)
+
+        word_weights = output.view(-1, self.ntokens)
+        word_idx = torch.multinomial(word_weights, 1)
+
+        return word_idx
+
+    def forward(self, X, training_behavior=False):
+        self.module_.train(training_behavior)
+
+        iterator = self.get_iterator(X, train=training_behavior)
+        y_probas = []
+        for x in iterator:
+            x = inferno.utils.to_var(x, use_cuda=self.use_cuda)
+            y_probas.append(self.evaluation_step(x))
+        return torch.cat(y_probas, dim=0)
+
+    def score(self, X, y):
+        # TODO: we cannot use predict() directly as the y supplied by GridSearchCV
+        # is not a "valid" y and only based on the input given to fit() down below.
+        # Therefore we have to generate our own batches.
```

Comment:
Is there an easy way to make it so that the provided `y` is actually valid? Do we maybe need a method that just applies the batching to `X` and/or `y`?